### PR TITLE
ci: benchmark stage in the jenkins pipeline

### PR DESCRIPTION
**Pull request by v1v**
fixes https://github.com/elastic/hey-apm/issues/119

## Highlights
- Enable benchmark stage in the Jenkins pipeline.
- First iteration using Docker Compose to prepare the context and run the benchmarking.

## Tasks
- [x] What data is required to be exported?
- [x] `-run 5m` doesn't see to work
- [x] Do we want to export the data to `https://apm-agent-benchmarks.app.elstc.co` or somewhere else?
- [x] Move `sendBenchmark` method to the apm pipeline library.

## Test cases
- [build](https://apm-ci.elastic.co/job/apm-server/job/apm-hey-test-mbp/view/change-requests/job/PR-125/7/) took 10 minutes to run in the bare-metal.
- [benchmark](https://apm-ci.elastic.co/job/apm-server/job/apm-hey-test-mbp/view/change-requests/job/PR-125/42/) took 38 minutes to run in the bare-metal using the docker-compose.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        -E apm-server.expvar.enabled=true
+        -E apm-server.frontend.enabled=true
+        -E apm-server.frontend.rate_limit=100000
+        -E apm-server.host=0.0.0.0:8200
+        -E apm-server.read_timeout=1m
+        -E apm-server.shutdown_timeout=2m
+        -E apm-server.write_timeout=1m
+        -E setup.template.settings.index.number_of_replicas=0
+        -E xpack.monitoring.elasticsearch=true
+        -E output.elasticsearch.enabled=${APM_SERVER_ELASTICSEARCH_OUTPUT_ENABLED:-true}
```
[the remark this answers is not in the document]
Comment (v1v) [reply docker-compose.yml]: I actually took that from https://github.com/elastic/opbeans-go/blob/master/docker-compose.yml#L58 so I'll remove it then.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    depends_on:
+      elasticsearch:
+        condition: service_healthy
+    healthcheck:
+      test: ["CMD", "curl", "--write-out", "'HTTP %{http_code}'", "--silent", "--output", "/dev/null", "http://apm-server:8200/healthcheck"]
+      retries: 10
+      interval: 10s
+
+  elasticsearch:
+    image: docker.elastic.co/elasticsearch/elasticsearch:${STACK_VERSION:-8.0.0-SNAPSHOT}
```
[the remark this answers is not in the document]
Comment (v1v) [reply docker-compose.yml]: Exactly, let's iterate firstly once using the docker-compose and then we can move forward with some follow-ups if required.

**Comment by v1v** [comment]
I'm able to reproduce the same issues locally:

```
vault read secret/apm-team/ci/apm-server-benchmark-cloud
STACK_VERSION=7.2.0 \
ES_URL="" \
ES_USER="" \
ES_PASS="" \
USER_ID="$(id -u):$(id -g)" \
docker-compose  up --no-color --exit-code-from hey-apm --remove-orphans --abort-on-container-exit --build hey-apm
```

```
hey-apm_1        | 2019/07/30 16:04:06 server.go:35: server status not OK: 404 Not Found
hey-apm_1        | 2019/07/30 16:04:06 run.go:106: apm-server version 7.2.0 built on 20 June [9a099b6]
hey-apm_1        | 2019/07/30 16:04:06 run.go:43: es-url unset: not indexing report
hey-apm_1        | cooling down 60.0 seconds... 
hey-apm_1        | running benchmark with transactions only
hey-apm_1        | 2019/07/30 16:05:06 server.go:35: server status not OK: 404 Not Found
hey-apm_1        | 2019/07/30 16:05:16 logger.go:10: [debug] sent request with 101 transactions, 0 spans, 0 errors, 0 metricsets
hey-apm_1        | 2019/07/30 16:05:26 logger.go:10: [debug] sent request with 101 transactions, 0 spans, 0 errors, 0 metricsets
hey-apm_1        | 2019/07/30 16:05:36 logger.go:10: [debug] sent request with 98 transactions, 0 spans, 0 errors, 0 metricsets
hey-apm_1        | 2019/07/30 16:05:36 run.go:36: 5.7862ms elapsed since event generation completed
hey-apm_1        | transactions sent ............. 300
hey-apm_1        | transactions dropped .......... 0
hey-apm_1        |  - success % .................. 100.00
hey-apm_1        | spans sent .................... 0
hey-apm_1        | spans dropped ................. 0
hey-apm_1        | spans sent per transaction .... 0.00
hey-apm_1        | errors sent ................... 0
hey-apm_1        | errors dropped ................ 0
hey-apm_1        | total events sent ............. 300
hey-apm_1        |  - per second ................. 10.00
hey-apm_1        |  - accepted ................... 300
hey-apm_1        | total requests ................ 3
hey-apm_1        | failed ........................ 0
hey-apm_1        | 2019/07/30 16:05:36 server.go:35: server status not OK: 404 Not Found
hey-apm_1        | 2019/07/30 16:05:36 run.go:106: apm-server version 7.2.0 built on 20 June [9a099b6]
hey-apm_1        | panic: runtime error: invalid memory address or nil pointer dereference
hey-apm_1        | [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x77e9d7]
hey-apm_1        | 
hey-apm_1        | goroutine 1 [running]:
hey-apm_1        | github.com/elastic/hey-apm/es.IndexReport(0xc00000d360, 0x7ffc7da37ebe, 0x9, 0x7ffc7da37ed1, 0xa, 0x7ffc7da37edc, 0xa, 0x7ffc7da37e9f, 0x16, 0x0, ...)
hey-apm_1        | 	/build/es/api.go:51 +0x207
hey-apm_1        | github.com/elastic/hey-apm/worker.Run(0x7ffc7da37e9f, 0x16, 0x0, 0x0, 0x7ffc7da37ebe, 0x9, 0x7ffc7da37ed1, 0x15, 0x7ffc7da37ef3, 0x19, ...)
hey-apm_1        | 	/build/worker/run.go:46 +0x6f6
hey-apm_1        | github.com/elastic/hey-apm/benchmark.runner.func1(0x86ffb3, 0x11, 0x7ffc7da37e9f, 0x16, 0x0, 0x0, 0x7ffc7da37ebe, 0x9, 0x7ffc7da37ed1, 0x15, ...)
hey-apm_1        | 	/build/benchmark/benchmark.go:74 +0x146
hey-apm_1        | github.com/elastic/hey-apm/benchmark.Run(0x7ffc7da37e9f, 0x16, 0x0, 0x0, 0x7ffc7da37ebe, 0x9, 0x7ffc7da37ed1, 0x15, 0x7ffc7da37ef3, 0x19, ...)
hey-apm_1        | 	/build/benchmark/benchmark.go:42 +0x26f
hey-apm_1        | main.main()
hey-apm_1        | 	/build/main.go:72 +0x98d
```

As far as I see there are a few weird issues:
- `2019/07/30 16:04:06 run.go:43: es-url unset: not indexing report`
- `2019/07/30 16:05:36 server.go:35: server status not OK: 404 Not Found`
- `panic: runtime error: invalid memory address or nil pointer dereference`

Not sure whether that's related with some memory issues in the docker containers or something else. But the first one regarding the `es-url` is not right, as it has been defined in the `CMD` to use the env variable `ES_URL`, which it does exist

```
docker inspect -f "{{ .Config.Env }}"  hey-apm_hey-apm_1 

[ES_URL=... ES_USER=... ES_PASS=... PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin]
```

**Comment by jalvz** [comment]
First issue: `es-url unset: not indexing report` that's very confusing, sorry!
It is because the tool sends a bunch of requests to apm-server to warm it up before starting the test, and those warm up requests are not indexed (on purpose). So if that messages appears only once, it's all good! 
I will change the logging :+1: 

Second issue `server status not OK: 404 Not Found`: that is because you need to start apm-server with `-E apm-server.expvar.enabled=true`. I'll change the log also for this

Third issue, the panic: good catch! 
will fix asap; but that happens because it couldn't save the benchmark results. we can't see the actual error because of the panic bug, but maybe something is off with ES_URL, or ES_PASS?

**Comment by v1v** [comment]
@jalvz , thanks. I'll enable `-E apm-server.expvar.enabled=true`. I'm still working on enabling to send the metrics to `-es-url` using `-es-auth`, for some reason the docker container does not interpolate some environment variables.

**Comment by jalvz** [comment]
Sounds good. If it is to complicated I can change the semantics so you pass `-es-user` and `-es-pwd` separately instead of concatenating them with `:`

**Comment by v1v** [comment]
I think I managed to resolve the interpolation issues with the CMD in the Dockefile by using the command step within the docker-compose.yml. See https://github.com/elastic/hey-apm/pull/125/commits/4ea5e3d1b7147806ad1bd42ec5d52d7e1e718d78

File: scripts/jenkins/run-bench-in-docker.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+USER_ID="$(id -u):$(id -g)" docker-compose \
+  up \
+  --no-color \
+  --exit-code-from hey-apm \
+  --build \
+  --remove-orphans \
+  --abort-on-container-exit \
+  hey-apm
+
+docker-compose down -v
```
Comment (jalvz) [inline scripts/jenkins/run-bench-in-docker.sh]: not-very-important-question: will `down -v` still be executed if I interrupt the job (eg. SIGTERM)?

Comment (v1v) [reply scripts/jenkins/run-bench-in-docker.sh]: Good catch!

**Pull request merged**